File: apps/nursery_app/lib/core/password_rules.dart
/// Matches API RegisterRequest / ResetPasswordRequest Password rules.
abstract final class PasswordRules {
  static const hint =
      'Use 8+ characters with upper & lower case letters and a number.';

  static bool hasMinLen(String password) => password.length >= 8;

  static bool hasLetter(String password) =>
      RegExp(r'[A-Za-z]').hasMatch(password);

  static bool hasMixedCase(String password) =>
      RegExp(r'[a-z]').hasMatch(password) &&
      RegExp(r'[A-Z]').hasMatch(password);

  static bool hasNumber(String password) => RegExp(r'\d').hasMatch(password);

  static bool isValid(String password) =>
      hasMinLen(password) &&
      hasLetter(password) &&
      hasMixedCase(password) &&
      hasNumber(password);

  static String? validate(String? value) {
    if (value == null || value.isEmpty) {
      return 'Enter a password.';
    }
    if (!isValid(value)) {
      return hint;
    }
    return null;
  }
}

File: apps/nursery-web/src/lib/password-rules.ts
/** Matches API RegisterRequest / ResetPasswordRequest Password rules. */
export function passwordRequirementErrors(password: string): string[] {
  const errors: string[] = [];
  if (password.length < 8) errors.push("At least 8 characters");
  if (!/[A-Za-z]/.test(password)) errors.push("At least one letter");
  if (!/[a-z]/.test(password) || !/[A-Z]/.test(password)) {
    errors.push("Upper and lower case letters");
  }
  if (!/\d/.test(password)) errors.push("At least one number");
  return errors;
}

export function isPasswordValid(password: string): boolean {
  return passwordRequirementErrors(password).length === 0;
}

export const PASSWORD_HINT =
  "Use 8+ characters with upper & lower case letters and a number.";
